src/cycax/cycad/engine.py

```diff
@@ -1,13 +1,217 @@
+            raise ValueError(msg)
+
+
+    def decode(self, data_file: str):
+        """
+        This is the main working class for decoding the scad. It is necessary for it to be refactored.
+        
+        !!!For this method to work properly it will be necessary to add a JSON, STL and SCAD file into the working repository.!!!
+        
+        Args: 
+            data_file : name of the file that is to be decoded into a scad.
+        """
+        out_name = "./SCAD/" + data_file + ".scad"
+        SCAD = open(out_name, "w")
+        in_name = "./JSON/" + data_file + ".json"
+        
+        with open(in_name) as f:
+            data = json.load(f)
+
+        output = []
+        dif = 0
+        for action in data:
+            if action["type"] == "cut":
+                dif = dif + 1
+                output.insert(0, self.decode_cut())
+
+            output.append(self.translate(action))
+
+
+            if action["side"] != None:
+                output.append(self.rotate(action["side"]))
+
+            if action["name"] == "cube":
+                output.append(self.decode_cube(action))
+
+            if action["name"] == "external":
+                output.append(self.decode_external(data_file))
+
+            if action["name"] == "hole":
+                output.append(self.decode_hole(action))
+
+            if action["name"] == "nut":
+                output.append(self.decode_nut(action))
+
+        i = 0
+        while i < dif:
+            i = i + 1
+            output.append("}")
+
+        for out in output:
+            SCAD.write(out)
+
+        SCAD.close()
+
+    def render_stl(self, filename: str):
+        """
+        
+        This takes a SCAD object and runs a command through linex that converts it into an stl. 
+        Depending on the complexity of the object it can take long to compute.
+        It prints out some messages to the terminal so that the impatient user will hopefully wait.(Similar to many windows request.)
+        
+        Args:
+            filename : This is the name of the file which will be converted from a scad to a stl
+        
+        """
+        out_name = "./SCAD/" + filename + ".scad"
+        out_stl_name = "./STL/" + filename + ".stl"
 
+        print("!!! THIS WILL TAKE SOME TIME, BE PATIENT !!!")
+        result = subprocess.run(
+            ["openscad", "-o", out_stl_name, out_name], capture_output=True, text=True
+        )
+        print("stdout:", result.stdout)
+        print("stderr:", result.stderr)
+        
     def add(self, part):
-        """This adds a new object into the assembly and decodes it into a json if that's what the user wants"""
-        jsonFile = open(part.part_no + ".json", "w")
+        """
+        This takes the provided part and will create its dictionary and export it to a json
+        
+        Args:
+            part(CycadPart) : This is the part that will be eported to a json.  
+        """
```

Comment:
A tip:
It is usually better to open a file in a context operator (with). e.g. https://www.freecodecamp.org/news/with-open-in-python-with-statement-syntax-example/ (I have not full read the article, but it looks ok). Writing JSON to file is such a common task that there is a method for it. [dump](https://docs.python.org/3/library/json.html#basic-usage).

The code then might look like:
```
with open("./JSON/" + part.part_no + ".json", "w") as fh:
   json.dump(part.export(), fh, indent=4)
```